python/cog/server/worker.py

```diff
@@ -44,44 +46,152 @@
+
+# def state_transition(old: WorkerState, new: WorkerState):
+#     def decorator(asyncgen: Callable[P, AsyncIterator[T]]) -> Callable[P, AsyncIterator[T]:
+#         @functools.wraps(asyncgen)
+#         def inner(self: "Worker", *args: P.args) -> AsyncIterator[T]:
+#             if self._state != old:
+#                 raise InvalidStateException(
+#                     f"Invalid operation: state is {self._state} (must be {old})"
+#                 )
+#             self._state = new
+#             return asyncgen(*args)
+
+
 class Worker:
-    def __init__(self, predictor_ref: str, tee_output: bool = True) -> None:
+    def __init__(
+        self, predictor_ref: str, tee_output: bool = True, concurrent: int = 1
+    ) -> None:
         self._state = WorkerState.NEW
         self._allow_cancel = False
+        self._semaphore = asyncio.Semaphore(concurrent)
+        self._concurrent = concurrent
 
         # A pipe with which to communicate with the child worker.
         events, child_events = _spawn.Pipe()
-        self._events: "AsyncPipe[_PublicEventType]" = AsyncPipe(events)
         self._child = _ChildWorker(predictor_ref, child_events, tee_output)
-        self._terminating = False
+        self._events: PipeType = AsyncPipe(events, self._child.is_alive)
+        # shutdown requested
+        self._shutting_down = False
+        # stop reading events
+        self._terminating = asyncio.Event()
+        self._mux = Mux(self._terminating)
+        self._predictions_in_flight = set()
+
+    # @contextlib.contextmanager
+    # def ctx(self):
+    #     self._events = AsyncPipe(events)
+    #     with self._events.executor:
+    #         self._child = _ChildWorker(predictor_ref, child_events, tee_output)
+    #         yield self
+    #         if self._child.is_alive():
+    #             self._child.terminate()
+    #             self._child.join()
 
     def setup(self) -> AsyncIterator[_PublicEventType]:
         self._assert_state(WorkerState.NEW)
         self._state = WorkerState.STARTING
-        self._child.start()
 
-        return self._wait(raise_on_error="Predictor errored during setup")
+        async def inner() -> AsyncIterator[_PublicEventType]:
+            self._child.start()
+            self._ensure_event_reader()
+            async for event in self._mux.read("SETUP", poll=0.1):
+                yield event
+                if isinstance(event, Done):
+                    if event.error:
+                        raise FatalWorkerException(
+                            "Predictor errored during setup: " + event.error_detail
+                        )
+                    self._state = WorkerState.IDLE
+                    self._allow_cancel = False
+
+        return inner()
+
+    @contextlib.asynccontextmanager
+    async def prediction_ctx(self, input: PredictionInput) -> AsyncIterator[None]:
+        async with self._semaphore:
+            self._state = WorkerState.PROCESSING
```

Comment:
something we could do here like 

```suggestion
            self._state = WorkerState.PROCESSING
            if self._semaphore._value == 0:
                self._state = WorkerState.BUSY
            else:
                self._state = WorkerState.READY
```